### PR TITLE
pinephone: add support for modem setup and audio routing.

### PR DESCRIPTION
**Pull request by jjsullivan5196**
This pull adds the necessary components for 4G modem usage, with the following use-cases in mind:

- Configure audio setup to and from modem
- Make and receive calls
- Send and read SMS
- Use mobile broadband/cellular data
- Wake-on-Modem (for calls and texts)

For now I've tested this with ModemManager+NetworkManager, but this also provides the groundwork for environments that use `ofono` (or just plain AT commands with `atinout` or `screen`, if you feel clever :^) )

To get started, you first need to run `pinephone_setup-modem` once as root to do the initial modem configuration. (chances are postmarketOS already did this if you booted the phone once, but this might need to be done again incase of potential data loss)

Some use-cases might require further configuration of software (eg specific to NetworkManager) but the setup script should do most of the needed work. Make sure modem is powered up by running `echo 1 > /sys/class/modem-power/modem-power/device/powered_blocking` Some examples of usage that I've tested below:

Read sms with `mmcli`
```sh
$ mmcli -s '/org/freedesktop/ModemManager1/SMS/2'
  -----------------------------------
  Content    |              number: '+5555555555'
             |                text: 'Test'
  -----------------------------------
  Properties |            PDU type: 'deliver'
             |               state: 'received'
             |             storage: 'me'
             |                smsc: '+919810051829'
             |           timestamp: '160808142935+05'
```

Sample script for answering calls, using ModemManager dbus interface. Audio is setup thru `alsaucm`
```sh
#!/bin/bash

MM_IFACE="org.freedesktop.ModemManager1"
MATCH_CALLS="type='signal',sender='${MM_IFACE}',interface='${MM_IFACE}.Modem.Voice',member='CallAdded'"

do_accept() {
	while read TYPE TIME SENDER HEADER
	do
		if [[ "$HEADER" = *"Modem.Voice"* ]]; then
			read TYPE _P OBJ
			OBJ=$(echo "${OBJ}" | tr -d '"')
			echo "Pickup ${OBJ}"

			alsaucm \
				open PinePhone \
				set _verb 'Voice Call' \
				set _enadev 'Mic' \
				set _enadev 'Earpiece'
			
			dbus-send \
				--system \
				--dest="${MM_IFACE}" \
				--print-reply \
				"${OBJ}" \
				"${MM_IFACE}.Call.Accept"
		else
			read
		fi
	done
}

dbus-monitor --system "${MATCH_CALLS}" | do_accept
```

Add a NetworkManager connection for LTE (sub `<operator_apn>` for your carrier's data APN)
```sh
$ nmcli c add type gsm ifname '*' con-name LTE apn <operator_apn>
```

For Wake-on-Modem: I tested by doing `loginctl suspend` from a sway session, then calling the phone. It wakes up within 2 rings.

fwiw: these examples are only good for making sure the hardware/software interface works. More work still needs to be done to get some telephony front-end software in the repo, like #24538 Sxmo also should work out-of-the box with the NetworkManager+ModemManager pair https://git.sr.ht/~mil/sxmo-docs

### CODE REVIEW

**Comment by ericonr** [comment]
What else is necessary here?

**Comment by jjsullivan5196** [comment]
@ericonr This will probably be blocked until me or someone else updates the pinephone kernel, since some changes have been made to the codec driver and its controls. After that I think all that's needed is to source some new ucm configurations. I don't want to say I'll get it done soon, but I don't think this should take much more work.

**Comment by jjsullivan5196** [comment]
@ericonr I tested this PR again on 5.10.12 and everything is still working. I think this can be merged, if no other quality issues are in the way.

File: srcpkgs/pinephone-base/template
Comment (the-maldridge) [inline srcpkgs/pinephone-base/template:10]: Hmm why did this change?  Usually the base-files metapackage should be public-domain.  If there are other files that need to be included under other licenses, you can make another package for them.

Comment (jjsullivan5196) [reply srcpkgs/pinephone-base/template:10]: Yeah, there's some scripts/configs I borrowed from postmarketos. I'll put them in a separate package then, maybe `pinephone-utils`?

Comment (the-maldridge) [reply srcpkgs/pinephone-base/template:10]: Yeah that sounds like a good name.

Comment (ericonr) [reply srcpkgs/pinephone-base/template:10]: Out of curiosity, could we have pmos as an upstream instead of vendoring the scripts in? (assuming you aren't changing them much).

Comment (jjsullivan5196) [reply srcpkgs/pinephone-base/template]: That might be possible, the scripts are sourced from their source packages repo, so I'd need to split them out somehow. However, I think any changes I made can be re-done in a ~10 line patch.

**Comment by jjsullivan5196** [comment]
@the-maldridge I split out the utilities into `pinephone-utils`, and @ericonr I'm sourcing postmarketos' stuff directly from pmaports now. If it looks a little hacky, I can revert back to vendoring the files.

**Review by the-maldridge — APPROVED** [review]

**Comment by MungFuSensei** [comment]
Anything holding this up? I have this on my pinephone with sxmo and it's working fine.

File: srcpkgs/pinephone-utils/template
Comment (ericonr) [inline srcpkgs/pinephone-utils/template:17]:
```suggestion
distfiles="https://gitlab.com/postmarketOS/pmaports/-/archive/${_commit_pmos}/pmaports-${_commit_pmos}.tar.gz?path=${_pmos_device}>pmos-scripts-${version}.tar.gz"
```

Otherwise local builds will use your cache and not update the checksum.

File: srcpkgs/pinephone-utils/template
Comment (ericonr) [inline srcpkgs/pinephone-utils/template:22]: vsed

File: srcpkgs/pinephone-utils/patches/vendor/modem-rules.patch
Comment (ericonr) [inline srcpkgs/pinephone-utils/patches/vendor/modem-rules.patch:8]: Can you explain the reason for the patch at the top?

File: srcpkgs/pinephone-utils/template
Comment (Piraty) [inline srcpkgs/pinephone-utils/template:15]: instead of the subshell, use `${_pmos_device//\//-}`

Comment (jjsullivan5196) [reply srcpkgs/pinephone-utils/template]: done :+1:

File: srcpkgs/pinephone-utils/template
Comment (jjsullivan5196) [reply srcpkgs/pinephone-utils/template]: yep :+1:

**Comment by the-maldridge** [comment]
Just an update on this.  Pretty much all available maintainer effort is going into the openssl migration next weekend, but I have my shiny new pinephone and I'm looking at testing this PR when I get time, then it will be merged.

File: srcpkgs/pinephone-utils/patches/vendor/modem-rules.patch
Comment (jjsullivan5196) [reply srcpkgs/pinephone-utils/patches/vendor/modem-rules.patch:8]: @ericonr Initially I took this out because I noticed this invocation didn't do anything, but that might be because the upstream script is 'broken' in a sense. The main parts of it are tested, but the variable which contains the path to the modem device file is unset https://gitlab.com/postmarketOS/pmaports/-/blob/master/device/main/device-pine64-pinephone/setup-modem.sh

This might have gone unnoticed, since the changes made by this script should be persistent until new settings are uploaded to the modem. I'll ask the pmos people about it, and try the script again myself. If nothing is wrong, I'll just kill this patch.

**Comment by JamiKettunen** [comment]
Is this still getting tested and merged one day? I'm getting a PinePhone later this month and would love to mess around on Void with it

**Comment by HadetTheUndying** [comment]
My existing pinephone is having some power on issues right now, but intend to pre-order the Pro and would love to be running Void on it. Is there anything I can do to help while I wait on the Pro to ship? They're supposed to send me some emails about figuring out what exactly is wrong with my pmOS edition Pinephone so I might be able to get it back up and running soon.

**Pull request closed without merging**
